index-scheduler/src/features.rs

```diff
@@ -73,9 +71,12 @@ impl FeatureData {
     pub fn new(env: &Env, instance_features: InstanceTogglableFeatures) -> Result<Self> {
         let mut wtxn = env.write_txn()?;
         let runtime_features = env.create_database(&mut wtxn, Some(EXPERIMENTAL_FEATURES))?;
+        let default_features =
+            RuntimeTogglableFeatures { metrics: instance_features.metrics, ..Default::default() };
+        runtime_features.put(&mut wtxn, EXPERIMENTAL_FEATURES, &default_features)?;
         wtxn.commit()?;
 
-        Ok(Self { runtime: runtime_features, instance: instance_features })
+        Ok(Self { runtime: runtime_features })
     }
```

Comment:
You are right! Good catch! However, after reviewing the different behaviors, it is likely our best behavior when combining the CLI arg and the runtime HTTP route. I would like to document and advice to always use the HTTP route and avoid using the CLI argument.

Reply:
Just wanted to check to make sure I'm understanding expected behavior correctly while working on updating this PR.

I just ran the following test:
#### Step One: Run server with `--experimental-enable-metrics`
```bash
cargo +nightly run -- --experimental-enable-metrics
```

#### Step Two: Check experimental features via HTTP
```bash
curl -X GET 'http://localhost:7700/experimental-features/'

{
  "scoreDetails": false,
  "vectorStore": false,
  "metrics": true
}
```

#### Step Three: Run server without `--experimental-enable-metrics`
```bash
cargo +nightly run
```

#### Step Four: Check experimental features via HTTP
```bash
curl -X GET 'http://localhost:7700/experimental-features/'

{
  "scoreDetails": false,
  "vectorStore": false,
  "metrics": false
}
```
---
Is this the expected behavior? Or am I missing something about my testing that would cause different behavior?

Reply:
Thank you for this information. Does it mean not specifying the flag will always put `"metrics": false` in the database?

Reply:
Hello @vivek-26, @braddotcoffee, @Kerollmops  :wave:

The complexity here is that we have two states to handle:

1. The "immediate" state of the feature, is it currently enabled or not? ("Feature state")
2. The "persistent" state stored in the DB ("DB state").

And two ways to act on the state:

1. Setting or not the CLI flag
2. Calling the experimental route to enable/disable the feature.

Taking this into account, I feel like the most intuitive behavior would be:

| DB state before operation | Action | Feature state after action | DB state after action | Note |
|---------------------------|--------|--------------------------|----------------------|------|
| ❌ | Start without CLI flag | ❌ | ❌ | |
| ❌ | Start with CLI flag | ✅ | ❌ | CLI flag enable features, doesn't change DB state |
| ❌ | Enable by calling the route | ✅ | ✅ | Regardless of the presence or absence of the CLI flag at startup |
| ❌ | Disable by calling the route | ❌ | ❌ | Regardless of the presence or absence of the CLI flag at startup |
| ✅ | Start without CLI flag | ✅ | ✅ | Lack of CLI flag never disables the feature, doesn't change DB state |
| ✅ | Start without CLI flag | ✅ | ✅ | |
| ✅ | Enable by calling the route | ✅ | ✅ | Regardless of the presence or absence of the CLI flag at startup |
| ✅ | Disable by calling the route | ❌ | ❌ | Regardless of the presence or absence of the CLI flag at startup |

Like @vivek-26 implied, setting or not setting the CLI flag should never affect the DB state, only the "immediate" feature state. Furthermore, not setting the CLI flag should not disable the feature, in that case the DB state should take precedence.

In short `initialState = db.state || cli.state`

Reply:
I can happily update to this! Just wanna make sure everyone is on the same page about it. @vivek-26 had mentioned in the original comment that this behavior might be unintuitive:
> let's say initially (when the server starts), at runtime level, metrics is false in DB, and, we pass `--experimental-enable-metrics` which gets persisted to DB.
>
> Now, if the user never updates the metrics feature using `PATCH /experimental-features`, **the next time if the server runs without `--experimental-enable-metrics`, the metrics feature will be enabled (as it was persisted to DB the last time server had started), which is incorrect I feel.**

This seems to be the behavior described by this line of your table:
| DB state before operation  | Action | Feature State After Action | DB State After Action | Note | 
| --- | --- | --- | --- | --- |
| ✅ | Start without CLI flag | ✅ | ✅ | Lack of CLI flag never disables the feature, doesn't change DB state |

@vivek-26 / @Kerollmops are you in alignment with @dureuill's table?